### PR TITLE
types(python): Azure DevOps Integration

### PR DESCRIPTION
**Pull request by mgaeta**
This PR adds types to `src/sentry/integrations/vsts/`. The code in this directory was pretty straightforward to type but I've pulled out some of the trickier bits into their own commits. For example:
- We use `UNSET = object()` as a placeholder for unset values so that we can still explicitly pass `None` as a parameter.
- We use the variable names `instance`, `base_url`, and `domain_name` interchangeably for an optional string which can be confusing

### CODE REVIEW

File: src/sentry/integrations/vsts/integration.py [hunk 1]
```diff
@@ -23,9 +23,9 @@
 )
 from sentry.integrations.repositories import RepositoryMixin
 from sentry.integrations.vsts.issues import VstsIssueSync
+from sentry.models import Identity
```
Comment (mgaeta) [inline src/sentry/integrations/vsts/integration.py:26]: y tho

File: src/sentry/integrations/vsts/client.py [hunk 2]
```diff
@@ -204,30 +225,31 @@ def get_commit_range(self, instance, repo_id, start_sha, end_sha):
...
+            # https://docs.microsoft.com/en-us/rest/api/azure/devops/core/projects/list?view=azure-devops-rest-6.1
+            # ). The token comes as a response header instead of the body and our API clients
+            # currently only return the body we can use count, $skip, and $top to get the same result.
             offset = self.page_size * page_number
             return {"stateFilter": "WellFormed", "$skip": offset, "$top": page_size}
 
-        def get_results(resp):
-            return resp["value"]
+        def get_results(resp: Response) -> Sequence[Any]:
+            return resp["value"]  # type: ignore
```
Comment (evanh) [inline src/sentry/integrations/vsts/client.py:244]: Why the ignore here? Could you not assert the type?

File: src/sentry/integrations/vsts/integration.py [hunk 3]
```diff
@@ -182,10 +187,10 @@ def get_organization_config(self):
                     for state in project_states:
                         all_states.add(state["name"])
 
-            all_states = [(state, state) for state in all_states]
+            all_states = {(state, state) for state in all_states}
             disabled = False
         except (ApiError, IdentityNotValid):
-            all_states = []
+            all_states = set()
```
Comment (evanh) [inline src/sentry/integrations/vsts/integration.py:193]: Why is this a set now?

File: src/sentry/integrations/vsts/integration.py [hunk 4]
```diff
@@ -370,15 +384,16 @@ def get_pipeline_views(self):
...
 
-    def build_integration(self, state):
+    def build_integration(self, state: Mapping[str, Any]) -> Mapping[str, Any]:
         data = state["identity"]["data"]
         oauth_data = self.get_oauth_data(data)
         account = state["account"]
         user = get_user_info(data["access_token"])
         scopes = sorted(self.get_scopes())
-        base_url = self.get_base_url(data["access_token"], account["accountId"])
+        # TODO(mgaeta): get_base_url() can return None.
```
Comment (evanh) [inline src/sentry/integrations/vsts/integration.py:393]: Shouldn't we handle that case?

**Review by evanh — COMMENTED** [review]
Couple questions, but mostly looks good.

File: src/sentry/integrations/vsts/client.py (hunk 2 above)
Comment (mgaeta) [reply src/sentry/integrations/vsts/client.py:244]: Good idea, I'll update this.

File: src/sentry/integrations/vsts/integration.py (hunk 3 above)
Comment (mgaeta) [reply src/sentry/integrations/vsts/integration.py:193]: I was trying to be consistent to the original type of `all_states`:
```python
try:
    all_states = set()
    for idx, project in enumerate(projects):
		...
        all_states.add(state["name"])
    all_states = [(state, state) for state in all_states]
except (ApiError, IdentityNotValid):
    all_states = []
```
but you're right it should end up as an array.

File: src/sentry/integrations/vsts/integration.py (hunk 4 above)
Comment (mgaeta) [reply src/sentry/integrations/vsts/integration.py:393]: This a tough one for me. `base_url` is used in a lot of places that just won't work with `None`. I think it's a bug in `get_base_url()` that `None` is a possible return value. I don't want to fix the bug in this PR so I'll propagate the optionality of the string all the way down and leave a TODO.

**Review by evanh — APPROVED** [review]
Approved, since the typing looks good and I can't see any functional changes. Someone on Ecosystem might want to take a scan just to make sure I didn't miss anything.

**Pull request merged**